### ISSUE TITLE
Helpers are emitted after imports

### ISSUE THREAD

**Issue by mhegazy** [Microsoft/TypeScript#2811]
Helpers are emitted at the top of the file before any imports. This way these helpers can not be imported using external module imports.

```ts
import 'reflect-metadata';
@inject
class House {
}
```
```js
var __metadata = this.__metadata || (typeof Reflect === "object" && Reflect.metadata) || function () { };
require('reflect-metadata');
```

that also applies to __extends, __decorate, __param as well

See https://github.com/Microsoft/TypeScript/issues/2249#issuecomment-93861855 for more details

**Comment by rbuckton** [comment]
There are several issues we would need to address:
1. The `__decorate` and `__metadata` helpers leverage a `Reflect` polyfill that may not yet have been imported, and need to be initialized inside of the module (or after any `require` calls)
2. For commonjs module emit, `this` inside the module is not the global.
3. For amd module emit, `this` inside the module may be null if the module body starts with a "use strict";
4. For ES6 emit, `this` inside of a module is `null`.

I have two proposals:

# Proposal 1
We leave the helpers before the module wrapper for ES5/ES3, but move the check for the `Reflect.decorate` and `Reflect.metadata` inside of the helpers. This will make each call to the helper more expensive, but will defer the execution until after the polyfill has loaded. Also, to fix the `this` binding in CommonJS and ES6, we would need to introduce a `__global` variable for the helper initializers to reference that would point to the global environment. 

Given the following source:
```ts
import "reflect-metadata";
export var x = 1;
```

You would have this output for `--module amd`:
```js
var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
var __extends = __global.__extends || ...;
var __decorate = __global.__decorate || ...;
define(["require", "exports", "reflect-metadata"], function (require, exports, reflectmetadata_1) {
  exports.x = 1;
});
```

# Proposal 2
We introduce a `__global` variable that our helper initializers would reference and helpers would be written inside the module body. 

Given the following source:
```ts
import "reflect-metadata";
export var x = 1;
```

You would have this output for `--module amd`:
```js
define(["require", "exports", "reflect-metadata"], function (require, exports, reflectmetadata_1) {
    var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
    var __extends = __global.__extends || ...;
    var __decorate = __global.__decorate || ...;
    var __metadata = __global.__metadata || ...;
    exports.x = 1;
});
```

You would have this output for `--module commonjs`:
```js
var reflectmetadata_1 = require("reflect-metadata");
var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
var __extends = __global.__extends || ...;
var __decorate = __global.__decorate || ...;
var __metadata = __global.__metadata || ...;
exports.x = 1;
```

You would have this output for `--module system`:
```js
System.register(["reflect-metadata"], function (exports_1) {
    var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
    var __extends = __global.__extends || ...;
    var __decorate = __global.__decorate || ...;
    var __metadata = __global.__metadata || ...;
    var x;
    return {
        setters: [...],
        execute: function() {
               exports_1("x", x = 1);
        }
    };  
});
```

And you would have this output for `--target ES6`
```js
import "reflect-metadata";
var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
var __extends = __global.__extends || ...;
var __decorate = __global.__decorate || ...;
var __metadata = __global.__metadata || ...;
export var x = 1;
```

# About `__global`
There are a number of ways of getting the global environment object, with differing downsides to each:

1. Testing with `typeof`
2. Using `eval`
3. Using `Function`

## Testing with `typeof`
One approach to detecting the global object is to use `typeof` testing to find the appropriate global based on the current execution environment:
```js
var __global = typeof global === "object" ? global : typeof self === "object" ? self : this;
```

This approach has the downside of being somewhat wordy, and requires that no other script has introduced a value named `global` or `self` with a different meaning. The `global` test is used for environments like nodejs/iojs that use the variable `global` (also `GLOBAL`) to refer to the global environment object. The `self` test is used for browsers, where `self` is the same as `window`, and web workers where `self` is the global environment for the worker. Finally, we fall back to `this` if we couldn't detect the global object.

## Using `eval`
Another, more concise approach is to use an unbound `eval` call, which will always execute in the global scope, and always executes outside of strict mode:
```js
var __global = (0, eval)("this");
```

The downsides to this approach are that it leverages a somewhat esoteric (but well documented) feature of the JavaScript language, and that can violate [CSP](http://www.w3.org/TR/CSP/).

## Using `Function`
As with **Using `eval`** above, you can use the `Function` constructor to create a function that executes outside of strict mode:
```js
var __global = Function("return this")();
```

This also has the same downsides as `eval`.

**Comment by rbuckton** [comment]
Or we do an even uglier thing, and just elide the `var`, and forcibly introduce a global (which we end up doing in AMD anyways):

```js
__extends = typeof __extends === "function" ? __extends : ...
__decorate = typeof __decorate === "function" ? __decorate : ...
```

**Comment by rbuckton** [comment]
Discussion relevant to `this` can now be found on #1622.

Instead of reordering the helper functions and imports, which could have side effects, we are deferring the check for `Reflect.decorate` and `Reflect.metadata` to the point at which the helper is called.

This is now fixed as of #2853.
